sound/soc/intel/boards/sof_maxim_common.c

```diff
@@ -137,6 +151,21 @@ int max_98373_spk_codec_init(struct snd_soc_pcm_runtime *rtd)
 	struct snd_soc_card *card = rtd->card;
 	int ret;
 
+	ret = snd_soc_dapm_new_controls(&card->dapm, maxim_2spk_widgets,
+					ARRAY_SIZE(maxim_2spk_widgets));
+	if (ret) {
+		dev_err(rtd->dev, "fail to add max98373 widgets, ret %d\n", ret);
+		return ret;
+	}
+
+	ret = snd_soc_add_card_controls(card, maxim_2spk_kcontrols,
+					ARRAY_SIZE(maxim_2spk_kcontrols));
+	if (ret) {
+		dev_err(rtd->dev, "fail to add max98373 kcontrols, ret %d\n",
+			ret);
+		return ret;
+	}
+
```

Comment:
Not fully following, shouldn't all this be depending on a num_codecs switch, e.g.

````
	switch (num_codecs) {
	case 2:
````

By moving the code from machine driver to a common part, we implicitly need to add a test, no?

Reply:
Machine driver registers 2 speaker widgets for those amplifiers so I didn't add extra code to check it. Sure I could add code to test the number of amplifier instance in another patch.

Reply:
probably best to check in the helpers that the number of amps/codecs is as expected. the code will also be more consistent this way.